runtime/argument_validation.go

```diff
@@ -0,0 +1,354 @@
+		_, ok = value.(interpreter.FunctionValue)
+	case interpreter.BlockDynamicType:
+		_, ok = value.(interpreter.BlockValue)
+	case interpreter.DeployedContractDynamicType:
+		_, ok = value.(interpreter.DeployedContractValue)
+	case interpreter.StorageReferenceDynamicType:
+		// Currently this only checks whether the the value is `interpreter.StorageReferenceValue`.
+		// It doesn't check whether the 'referenced' value conforms to the innerType.
+		// TODO: add support for checking the referenced value conformance (if importing is supported).
+		_, ok = value.(*interpreter.StorageReferenceValue)
+	case interpreter.EphemeralReferenceDynamicType:
+		ok = valueConformsToEphemeralReferenceDynamicType(value, typ)
+	case interpreter.VoidDynamicType:
+		// Void type cannot have a value.
+		ok = false
+	}
+
+	return
+}
+
+func valueConformsToArrayDynamicType(value interpreter.Value, arrayType interpreter.ArrayDynamicType) bool {
+	arrayValue, ok := value.(*interpreter.ArrayValue)
+	if !ok || len(arrayValue.Values) != len(arrayType.ElementTypes) {
+		return false
+	}
+
+	for index, item := range arrayValue.Values {
+		if !valueConformsToDynamicType(item, arrayType.ElementTypes[index]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func valueConformsToDictionaryDynamicType(
+	value interpreter.Value,
+	dictionaryType interpreter.DictionaryDynamicType,
+) bool {
+
+	dictionaryValue, ok := value.(*interpreter.DictionaryValue)
+	if !ok || len(dictionaryValue.Keys.Values) != len(dictionaryType.EntryTypes) {
+		return false
+	}
+
+	for index, entryKey := range dictionaryValue.Keys.Values {
+		entryType := dictionaryType.EntryTypes[index]
+
+		// Check the key
+		if !valueConformsToDynamicType(entryKey, entryType.KeyType) {
+			return false
+		}
+
+		// Check the value. Here it is assumed an imported value can only have
+		// static entries, but not deferred keys/values.
+		key := interpreter.DictionaryKey(entryKey)
+		entryValue, ok := dictionaryValue.Entries.Get(key)
+		if !ok || !valueConformsToDynamicType(entryValue, entryType.ValueType) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func valueConformsToSomeDynamicType(value interpreter.Value, someType interpreter.SomeDynamicType) bool {
+	someValue, ok := value.(*interpreter.SomeValue)
+	if !ok {
+		return false
+	}
+
+	return valueConformsToDynamicType(someValue.Value, someType.InnerType)
+}
+
+func valueConformsToEphemeralReferenceDynamicType(
+	value interpreter.Value,
+	refType interpreter.EphemeralReferenceDynamicType,
+) bool {
+
+	referenceValue, ok := value.(*interpreter.EphemeralReferenceValue)
+	if !ok {
+		return false
+	}
+
+	return valueConformsToDynamicType(*referenceValue.ReferencedValue(), refType.InnerType())
+}
+
+// valueConformsToSemaType checks whether a value conforms to a given semantic type.
+// Returns `true` if the value and the type matches, and `false` otherwise.
+//
+func valueConformsToSemaType(value interpreter.Value, semaType sema.Type) (ok bool) {
```

Comment:
Just as mentioned above this could be maybe a function, potentially inverted, i.e on `interpreter.Value` and taking a `sema.Type` parameter, instead of the other way round (on `sema.Type`) because that would lead to cyclic imports of packages

Reply:
added with 36327c9